File: ws_client_dart/bin/transport/message_handler.dart
import 'dart:convert';
import 'message_type.dart';

// Para transformar el mensaje si se envia o se recibe
class MessageHandler {
  // Método para crear un mensaje JSON, Para enviar
  String encodeMessageJson(MessageType type, dynamic object) {
    return jsonEncode({
      'type': type.toString().split(".").last,
      'object': object is String
          ? object
          : object.toJson(), // Convertimos a String si es necesario
    });
  }

// Método para manejar un mensaje JSON, Para recibir
  Map<String, dynamic> decodeMessageJson(String message) {
    // Paso 1: Convertir el mensaje JSON en un Map
    final Map<String, dynamic> jsonData = jsonDecode(message);

    // Paso 2: Leer el tipo de mensaje y convertirlo de nuevo al enum
    final String typeString = jsonData['type'];
    final MessageType type = MessageType.values.firstWhere(
      (e) => e.toString().split('.').last == typeString,
      orElse: () => MessageType
          .unknow, // Valor por defecto en caso de no encontrar el tipo
    );

    // Paso 3: Leer el objeto
    final object = jsonData['object'];

    // Retornar el tipo como enum y el objeto
    return {
      'type': type,
      'object': object,
    };
  }
}

File: ws_server_node/src/transport/MessageHandler.js
const MessageType = require('./MessageType');

class MessageHandler {
    // Método para crear un mensaje JSON, Para enviar
    encodeMessageJson(type, object) {
        return JSON.stringify({
            type: type,
            object: typeof object === 'string' ? object : object.toJson()
        });
    }

    // Método para manejar un mensaje JSON, Para recibir
    decodeMessageJson(message) {
        // Paso 1: Convertir el mensaje JSON en un objeto
        const jsonData = JSON.parse(message);

        // Paso 2: Leer el tipo de mensaje
        const typeString = jsonData.type;
        const type = Object.values(MessageType).includes(typeString)
            ? typeString
            : MessageType.UNKNOW;

        // Paso 3: Leer el objeto
        const object = jsonData.object;

        // Retornar el tipo y el objeto
        return {
            type: type,
            object: object
        };
    }
}

module.exports = MessageHandler;
